capi_problems.rst

```diff
@@ -70,7 +70,145 @@ and the particular requirements that each of them has.
 C API Stakeholders
 ==================
 
-[TODO: write this section]
+As mentioned in the introduction, the C API was originally
+created as the internal interface between CPython's
+interpreter and the Python layer. It was later exposed as
+a way for third party developers to extend and embed Python
+programs. Over the years, new types of stakeholders emerged,
+with different requirements and areas of focus. This section
+describes this complex state of affairs in terms of the
+actions that different stakeholders need to perform through
+the C API.
+
+There are actions which are generic, and required by
+all types of API users:
+
+* Define functions and call them
+* Create classes and instantiate them
+* Create instances of builtin and user defined types
+  and perform operations on them
+* Introspect objects, including types, instances, and functions.
+* Raise and handle exceptions
+* Import modules
+* Manage threads
+* Manage sub-interpreters
+* Handle and send signals
+
+Groups of users are:
+
+**Extension writers**
+
+These are the traditional users of the C API, and their requirements
+are listed above.
```

Comment:
Would it make sense to expand on other than functional aspects of the API? Extension users may have different requirements regarding stability and necessity to rebuild from sources for different Python versions and interpreters (backwards: will extension compiled on 3.10 work on 3.8, and forwards: will extension compiled today work on 3.15?).

CPython developers, which is maybe a stakeholder that's worth mentioning here (as long as we have not agreed on completely separating API for "the internal interface between CPython's interpreter" and API for Python extensions), do not need to care about stability that much and do not need to care about rebuilding at all, because internal code is rebuilt for each new version by definition. Another difference is the CPython internal code will never be meant to run on alternative Python wheres extension author may see that as an advantage.

Reply:
Thinking about it more, I think the CPython developers section is more than desirable because one of the IMO major issues of the current C API is that the compatibility requirements are holding CPython back from changing its internal implementation. 

The [Let’s get rid of the stable ABI, but keep the limited API](https://discuss.python.org/t/lets-get-rid-of-the-stable-abi-but-keep-the-limited-api/18458) threads comes to mind.

Another example: significant part of the current and the past GIL-removal attempts has been to work around scalability issues with reference counting. And they had to deal with that partly also because reference counting is exposed by the C API, otherwise alternative memory management approach could have been a viable solution.

Reply:
Yes, good points. Let me merge this and then you can make another PR.